spring-cloud-gcp-data-datastore/src/main/java/org/springframework/cloud/gcp/data/datastore/core/DatastoreTemplate.java

```diff
@@ -113,21 +117,42 @@ public DatastoreEntityConverter getDatastoreEntityConverter() {
 
 	@Override
 	public <T> T findById(Object id, Class<T> entityClass) {
-		Iterator<T> results = findAllById(Collections.singleton(id), entityClass)
-				.iterator();
+		return findById(id, entityClass, new ReadContext());
+	}
+
+	private  <T> T findById(Object id, Class<T> entityClass, ReadContext context) {
+		Iterator<T> results = findAllById(Collections.singleton(id), entityClass, context).iterator();
 		return results.hasNext() ? results.next() : null;
 	}
 
 	@Override
 	public <T> T save(T instance, Key... ancestors) {
-		getDatastoreReadWriter().put(convertToEntityForSave(instance, ancestors));
-		return instance;
+		return save(instance, new HashSet<>(), ancestors);
+	}
+
+	private <T> T save(T instance, Set<Key> persisted, Key... ancestors) {
+		Iterable<T> iterable = saveAll(Collections.singletonList(instance), persisted, ancestors);
+		return iterable.iterator().next();
 	}
 
 	@Override
 	public <T> Iterable<T> saveAll(Iterable<T> entities, Key... ancestors) {
-		getDatastoreReadWriter().put(StreamSupport.stream(entities.spliterator(), false)
-				.map((entity) -> convertToEntityForSave(entity, ancestors)).toArray(Entity[]::new));
+		return saveAll(entities, new HashSet<>(), ancestors);
+	}
+
+	private <T> Iterable<T> saveAll(Iterable<T> entities, Set<Key> persisted, Key... ancestors) {
+		List<Entity> entitiesForSave = new ArrayList<>();
+		for (T entity : entities) {
+			Key key = getKey(entity, true, ancestors);
+			if (persisted.contains(key)) {
+				continue;
+			}
+			persisted.add(key);
+			entitiesForSave.add(convertToEntityForSave(entity, persisted, ancestors));
+		}
+		if (!entitiesForSave.isEmpty()) {
+			getDatastoreReadWriter().put(entitiesForSave.toArray(new Entity[0]));
```

Comment:
Why not avoid this whole conversion from List to array?
```java
		Entity[] entitiesForSave = StreamSupport.stream(entities.spliterator(), false)
				.filter(persisted::contains)
				.peek(entity -> persisted.add(getKey(entity, true, ancestors)))
				.map(entity -> convertToEntityForSave(entity, persisted, ancestors))
				.toArray(Entity[]::new);
		if (entitiesForSave.length > 0) {
			getDatastoreReadWriter().put(entitiesForSave);
		}
		return entities;
```

Reply:
It looks like in both cases you need to do a `.toArray(Entity[]::new);` somewhere in there

Reply:
I guess you then have to do something like this:
```java
	Entity[] entitiesForSave = StreamSupport.stream(entities.spliterator(), false)
				.map(entity -> new AbstractMap.SimpleEntry<Key, T>(getKey(entity, true, ancestors), entity))
				.filter(entry -> persisted.contains(entry.getKey()))
				.peek(entry -> persisted.add(entry.getKey()))
				.map(entry -> convertToEntityForSave(entry.getValue(), persisted, ancestors))
				.toArray(Entity[]::new);
		if (entitiesForSave.length > 0) {
			getDatastoreReadWriter().put(entitiesForSave);
		}
		return entities;
```
A bit more ugly, but yeah.

Reply:
In this case I think it looks less ugly with a loop. The overhead of using ArrayList is not big, probably comparable with the stream overhead.

Reply:
It looks like 2x working memory overhead with the non-streaming version.